cmd/redirectserver/app/handlers.go

```diff
@@ -0,0 +1,228 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"net/http"
+	"os"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"k8s.io/klog/v2"
+
+	"k8s.io/registry.k8s.io/pkg/clientip"
+	"k8s.io/registry.k8s.io/pkg/net/cidrs"
+	"k8s.io/registry.k8s.io/pkg/net/cidrs/aws"
+
+	"sigs.k8s.io/porche/cmd/redirectserver/pkg/blobcache"
+)
+
+type MirrorConfig struct {
+	CanonicalLocation GCSKey
+	// CanonicalFallback is the fallback URL direct to a public bucket or similar,
+	// this is used when we're having problems finding a redirect bucket.
+	CanonicalFallback string
+
+	InfoURL    string
+	PrivacyURL string
+}
+
+type Server struct {
+	MirrorConfig
+
+	// contentCache stores the content of files, in particular the hash files
+	contentCache *contentCache
+
+	// regionMapper maps from a client IP to a cloud & region
+	regionMapper cidrs.IPMapper[string]
+
+	// mirrorCache records whether the mirrors have the various files.
+	// We use a lookup so we don't require 100% perfect replication to all mirrors.
+	mirrorCache blobcache.BlobChecker
+}
+
+func NewServer(cfg MirrorConfig, storageClient *storage.Client, mirrorCache blobcache.BlobChecker) *Server {
+	s := &Server{
+		MirrorConfig: cfg,
+	}
+	// cache of hash files
+	s.contentCache = newContentCache(storageClient)
+	// initialize map of clientIP to AWS region
+	s.regionMapper = aws.NewAWSRegionMapper()
+	// cache of whether the mirrors have blobs
+	s.mirrorCache = mirrorCache
+
+	return s
+}
+
+// ServeHTTP is the main entry point
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// only allow GET, HEAD; we don't allow
+	if r.Method != http.MethodGet && r.Method != http.MethodHead {
+		http.Error(w, "Only GET and HEAD are allowed.", http.StatusMethodNotAllowed)
+		return
+	}
+	path := r.URL.Path
+	switch {
+	case strings.HasPrefix(path, "/binaries/"):
+		s.serveBinaries(w, r)
+	case path == "/":
+		http.Redirect(w, r, s.InfoURL, http.StatusTemporaryRedirect)
+	case strings.HasPrefix(path, "/privacy"):
+		http.Redirect(w, r, s.PrivacyURL, http.StatusTemporaryRedirect)
+	default:
+		klog.V(2).InfoS("unknown request", "path", path)
+		http.NotFound(w, r)
+	}
+}
+
+func (s *Server) serveBinaries(w http.ResponseWriter, r *http.Request) {
+	rPath := r.URL.Path
+
+	serveDirectly := false
+
+	suffix := getSuffix(rPath)
+	switch strings.ToLower(suffix) {
+	case ".sha1", ".sha224", ".sha256", ".sha384", ".sha512", ".md5":
+		// Hashes are not redirected, both for security and because they're so small
+		serveDirectly = true
+	case ".gpg":
+		// Like hashes, we serve small security-critical key/signature information directly
+		serveDirectly = true
+	}
+
+	if serveDirectly {
+		s.serveDirectly(w, r)
```

Comment:
This is cute but I don't really think we should do this, we need to keep cloud run cost as small as possible, and at the end of the day we're still trusting https and just providing an integrity check file.

we *need* to offload costs here.

Reply:
So this is small enough that it's actually cheaper than a redirect :-)

However, we can't 302 redirect to a non-authoritative bucket without changing the security posture here.  The idea is that as a client, I choose to trust `artifacts.k8s.io` (or whatever it is), and as long as I check hash files, I don't need to trust anything else.  If we 302 redirect the hash files to `something.example.com`, now I as as user need to trust `something.example.com`.

Reply:
> So this is small enough that it's actually cheaper than a redirect :-)

Not true, because we're paying for the compute time to go fetch it, the cloud run model is we pay by compute time. This also adds unnecessary runtime complexity that we need to watch out for memory leaks etc.

> However, we can't 302 redirect to a non-authoritative bucket without changing the security posture here. The idea is that as a client, I choose to trust artifacts.k8s.io (or whatever it is), and as long as I check hash files, I don't need to trust anything else. If we 302 redirect the hash files to something.example.com, now I as as user need to trust something.example.com.

This is a false sense of security. If we want to give clients a stronger security posture we should sign releases.
Ultimately we're trusting the origin. We should not pretend that is not the case.

Reply:
As we discussed, I'll look into how we should sign releases.

I also didn't realize CloudRun was that expensive!

Reply:
I'll also pull out the direct serving :-)